### DISCUSSION TITLE
Build using edge compatible versions of libraries

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [sveltejs/kit#10456] [Q&A]
There are several npm modules I've tried to use with the Vercel adapter on the edge that result in build time errors. Each of these support the edge runtime, but for some reason Sveltekit keeps building using the NodeJS versions of the libraries.

**How can I tell Sveltekit to build using the edge version of the libraries?**

For context, here's my svelte config:

```js
import adapter from '@sveltejs/adapter-vercel'
import { vitePreprocess } from '@sveltejs/kit/vite'

/** @type {import('@sveltejs/kit').Config} */
const config = {
    preprocess: vitePreprocess(),

    kit: {
        adapter: adapter({
            runtime: 'edge',
        }),
    }
}
```

## Example: `jose`

I'll use Jose as an example because that's a fairly popular library.

When I build using Jose, I get the following error:

```
      3 │ import { Buffer } from "buffer";
        ╵                        ~~~~~~~~

  The package "buffer" wasn't found on the file system but is built into node. Are you
  trying to bundle for node? You can use "platform: 'node'" to do that, which will remove
  this error.
```

I also get the same errors for crypto, util, and zlib.

Jose [supports Vercel's edge](https://github.com/panva/jose/issues/301).

Interestingly, these errors seem to come after the main build step:

```
Run npm run preview to preview your production build locally.

> Using @sveltejs/adapter-vercel
✘ [ERROR] Could not resolve "buffer"

    .svelte-kit/output/server/chunks/auth.js:3:23:
      3 │ import { Buffer } from "buffer";
        ╵                        ~~~~~~~~

  The package "buffer" wasn't found on the file system but is built into node. Are you
  trying to bundle for node? You can use "platform: 'node'" to do that, which will remove
  this error.
```

## References

- Vercel Edge authentication example using Jose: https://github.com/vercel/examples/tree/main/edge-middleware/jwt-authentication
- A similar issue with Vite: https://github.com/vitejs/vite/issues/3772

**Comment by justinfriebel** [comment]
I had a similar issue but it was happening with libsql. I fixed it by importing the correct non-node version. I think your only option is to find if there is a proper non-node import from the packages causing issues. If it helps at all I switched to using Lucia auth because it seemed more edge friendly: [https://github.com/justinfriebel/sveltekit-turso-drizzle](https://github.com/justinfriebel/sveltekit-turso-drizzle)

**Reply by thenbe** [reply]
It'd be great if there was a quick way to know which dependency is causing the issue. The logs and error trace do not provide this information.

<details><summary>Logs and Error trace</summary>
<p>

```
...

Run npm run preview to preview your production build locally.

> Using @sveltejs/adapter-vercel
✘ [ERROR] Could not resolve "buffer"

    ../../node_modules/.pnpm/safe-buffer@5.2.1/node_modules/safe-buffer/index.js:3:21:
      3 │ var buffer = require('buffer')
        ╵                      ~~~~~~~~

  The package "buffer" wasn't found on the file system but is built into node. Are you trying to
  bundle for node? You can use "platform: 'node'" to do that, which will remove this error.

error during build:
Error: Build failed with 1 error:
../../node_modules/.pnpm/safe-buffer@5.2.1/node_modules/safe-buffer/index.js:3:21: ERROR: Could not resolve "buffer"
    at failureErrorWithLog (/home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:1649:15)
    at /home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:1058:25
    at /home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:1003:52
    at buildResponseToResult (/home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:1056:7)
    at /home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:1085:16
    at responseCallbacks.<computed> (/home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:703:9)
    at handleIncomingPacket (/home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:762:9)
    at Socket.readFromStdout (/home/nbe/projects/site-parent/worktrees/edge/node_modules/.pnpm/esbuild@0.18.20/node_modules/esbuild/lib/main.js:679:7)
    at Socket.emit (node:events:517:28)
    at addChunk (node:internal/streams/readable:335:12)
```

</p>
</details> 

Tried searching but `rg -uuu 'safe.*buffer'` came up with a lot of hits.

Anyways, in my case it was the [Feed](https://www.npmjs.com/package/feed) package causing the error. I fixed it by adding the following to my rss feed route:

```ts
export const config = {
  runtime: 'nodejs18.x', // `feed` package is not compatible with Vercel's edge runtime
}
```
